### PR TITLE
Vector gradient

### PR DESCRIPTION
**Pull request by fsih**
gradients
(Thanks MysticUnicornSparkle for project 144885018!)

### Resolves
Fixes https://github.com/LLK/scratch-paint/issues/541
Fixes https://github.com/LLK/scratch-paint/issues/362

### Proposed Changes
Adds gradient UI

### Testing
Up at https://llk.github.io/scratch-gui/gradients/

### CODE REVIEW

**Comment by fsih** [comment]
Sorry about dumping this on you Paul 🙏orz I'm going on vacation on Wednesday and really wanted to get this in. Hoping to get a group of people to bash on it on Monday so I can try to fix anything that comes up on Tuesday. Then you could take your time to review before the end of the week, and I can respond remotely, but I probably won't be able to fix big things at that point. Does that plan sound ok? 💦💦

File: src/components/color-picker/color-picker.jsx [hunk 1]
```diff
@@ -143,14 +278,26 @@ class ColorPickerComponent extends React.Component {
...
+    onSelectColor2: PropTypes.func.isRequired,
+    onSwap: PropTypes.func,
     onTransparent: PropTypes.func.isRequired,
-    saturation: PropTypes.number.isRequired
+    saturation: PropTypes.number.isRequired,
+    shouldShowGradientTools: PropTypes.bool.isRequired
 };
 
-export default ColorPickerComponent;
+export default connect()(injectIntl(ColorPickerComponent));
```
Comment (paulkaplan) [inline src/components/color-picker/color-picker.jsx]: I do not think connect should be required here

File: src/components/forms/slider.jsx [hunk 2]
```diff
@@ -63,7 +64,7 @@ class SliderComponent extends React.Component {
             halfHandleWidth;
         return (
             <div
-                className={styles.container}
+                className={this.props.lastSlider ? classNames(styles.container, styles.last) : styles.container}
```
Comment (paulkaplan) [inline src/components/forms/slider.jsx]: I think you can use:

```
className={classNames({
    [styles.container]: true,
    [styles.last]: this.props.lastSlider
})}
```

**Comment by fsih** [comment]
Fixes since smoke test (we might want to test these fixes):
- Fix not being able to move shapes with both gradients and holes
- Update icons, so that the "dark" color in the icon matches where color #1 will go
- Replace randomly generated color with color hue shifted by a consistent amount. This color appears  when you switch from a solid color to a gradient.
- Switch selection to second swatch when changing from solid to gradient
- Update paper, bringing in Paul's fix for exported text gradient locations

File: src/components/color-picker/color-picker.jsx (hunk 1 above)
Comment (fsih) [reply src/components/color-picker/color-picker.jsx]: fixed

File: src/helper/style-path.js [hunk 3]
```diff
@@ -30,31 +33,203 @@ const _getColorStateListeners = function (textEditTargetId) {
...
+ */
+const applyGradientTypeToSelection = function (gradientType, bitmapMode, textEditTargetId) {
+    const items = _getColorStateListeners(textEditTargetId);
+    let changed = false;
+    for (let item of items) {
+        if (item.parent instanceof paper.CompoundPath) {
+            item = item.parent;
+        }
+
+        let itemColor1 = item.fillColor === null || item.fillColor.alpha === 0 ? null :
```
Comment (paulkaplan) [inline src/helper/style-path.js]: Hmmmmmm if statements plz

Comment (fsih) [reply src/helper/style-path.js]: done

**Review by paulkaplan — APPROVED** [review]

**Pull request merged**